### PR TITLE
Quick fix in update fine search candidate for map accessors

### PR DESCRIPTION
**Pull request by acdaigneault**
# Description of the problem

- Weird crashes (segfault) in the `update_fine_search_candidates()` function related to the particle-particle containers
- Seems related to load balancing or the contact threshold or neither of that
- We still do not understand why the changes fixed the crashes and we are not sure it will prevent other crashes or it just fixed our edge case.
- Our hypothesis right now should be that the issues came from the way the particle contact data structure was not being cleared correctly after load balancing or particle location/distribution.

# Description of the solution

- Change accessor from [] to .find() for the candidate containers (broad search)

# How Has This Been Tested?

- Tests are good

### CODE REVIEW

**Review by blaisb — APPROVED** [review]

**Review by OGaboriault — APPROVED** [review]

**Comment by blaisb** [comment]
@OGaboriault Tell me when you think this is ready for merge
I think this is good, so most likely I will merge it tomorrow or smthing

**Comment by acdaigneault** [comment]
I change a few [] in the broad search also

File: include/dem/particle_wall_broad_search.h [hunk 1]
```diff
@@ -164,6 +164,90 @@ class ParticleWallBroadSearch
...
+    auto candidates_container_it = contact_pair_candidates.find(particle_id);
+
+    // Reserve arbitrary vector capacity and store if the particle does not have
+    // contact candidate yet
+    if (candidates_container_it == contact_pair_candidates.end())
+      {
+        auto pair_it_bool = contact_pair_candidates.emplace(
+          particle_id,
+          std::unordered_map<DEM::global_face_id,
+                             Particles::ParticleIterator<dim>>());
```
Comment (blaisb) [inline include/dem/particle_wall_broad_search.h:242]: Do you know if there's any reason why we still use an unordered map here?

**Review by blaisb — APPROVED** [review]
Should be good. I will try to stress test it on my machine to identify the consequences of this.

File: include/dem/particle_wall_broad_search.h (hunk 1 above)
Comment (acdaigneault) [reply include/dem/particle_wall_broad_search.h:242]: I think you have not change the small maps to ankerl

Comment (OGaboriault) [reply include/dem/particle_wall_broad_search.h:242]: Performance boost incoming?

File: source/dem/particle_particle_fine_search.cc [hunk 2]
```diff
@@ -7,11 +7,12 @@ template <int dim>
 ParticleParticleFineSearch<dim>::ParticleParticleFineSearch()
 {}
 
+//TODO PUT CONST
```
Comment (acdaigneault) [inline source/dem/particle_particle_fine_search.cc]: À enlever?

File: source/dem/update_local_particle_containers.cc [hunk 3]
```diff
@@ -59,10 +58,20 @@ update_contact_container_iterators(
           // Get current particle id
           unsigned int particle_id = pairs_in_contact_iterator->first;
 
+          // Find if particle one is still a particle on this process
```
Comment (acdaigneault) [inline source/dem/update_local_particle_containers.cc:63]: Find if particle one is still on this process

File: source/dem/update_local_particle_containers.cc [hunk 4]
```diff
@@ -76,27 +85,38 @@ update_contact_container_iterators(
...
                                 ghost_local_periodic_particle_particle)
                 {
-                  // For particle-particle contacts, iterators of both particles
-                  // must be updated
-                  adjacent_map_iterator->second.particle_one =
-                    particle_container[particle_id];
-
                   unsigned int particle_two_id = adjacent_map_iterator->first;
+
+                  // Find if particle two is still a particle on this process
```
Comment (acdaigneault) [inline source/dem/update_local_particle_containers.cc:93]: same

**Review by acdaigneault — COMMENTED** [review]
yay

File: source/dem/update_local_particle_containers.cc [hunk 5]
```diff
@@ -59,10 +60,21 @@ update_contact_container_iterators(
           // Get current particle id
           unsigned int particle_id = pairs_in_contact_iterator->first;
 
+          // Find if particle one is still a particle on this process
```
Comment (blaisb) [inline source/dem/update_local_particle_containers.cc:63]:
```suggestion
          // Find if particle one is still on this process
```

File: source/dem/update_local_particle_containers.cc [hunk 6]
```diff
@@ -76,27 +88,42 @@ update_contact_container_iterators(
...
                                 ghost_local_periodic_particle_particle)
                 {
-                  // For particle-particle contacts, iterators of both particles
-                  // must be updated
-                  adjacent_map_iterator->second.particle_one =
-                    particle_container[particle_id];
-
                   unsigned int particle_two_id = adjacent_map_iterator->first;
+
+                  // Find if particle two is still a particle on this process
```
Comment (blaisb) [inline source/dem/update_local_particle_containers.cc:93]:
```suggestion
                  // Find if particle two is still on this process
```

File: source/dem/particle_particle_fine_search.cc [hunk 7]
```diff
@@ -7,10 +7,11 @@ template <int dim>
 ParticleParticleFineSearch<dim>::ParticleParticleFineSearch()
 {}
 
+// TODO PUT CONST
```
Comment (blaisb) [inline source/dem/particle_particle_fine_search.cc:10]:
```suggestion
```

Comment (OGaboriault) [inline source/dem/particle_particle_fine_search.cc]:
```suggestion

```

**Review by OGaboriault — APPROVED** [review]

**Pull request merged**